src/react-vrpc.js

```diff
@@ -163,37 +169,106 @@ class VrpcBackendMaker extends Component {
           default:
             obj[key].push({ ...proxy, [event]: args })
         }
-        this.setState({ [key]: obj[key] })
+        this.setState({ [key]: { ...obj } })
       })
     }
   }
 
+  _renderProviders (children, index = 0) {
+    const Context = vrpcContexts[index]
+    const Provider = Context.Provider
+    if (index === 0) {
+      return (
+        <vrpcGlobalContext.Provider value={this.state.__global__}>
+          <Provider value={this.state[Context.displayName]}>
+            {this._renderProviders(children, index + 1)}
+          </Provider>
+        </vrpcGlobalContext.Provider>
+      )
+    }
+    if (index < vrpcContexts.length - 1) {
+      return (
+        <Provider value={this.state[Context.displayName]}>
+          {this._renderProviders(children, index + 1)}
+        </Provider>
+      )
+    }
+    return (
+      <Provider value={this.state[Context.displayName]}>
+        {children}
+      </Provider>
+    )
+  }
+
   render () {
     const { loading } = this.props
-    const { vrpcIsLoading } = this.state
+    const { vrpcIsLoading } = this.state.__global__
     if (vrpcIsLoading) return loading || false
     const { children } = this.props
-    return (
-      <VrpcContext.Provider value={this.state}>
-        {children}
-      </VrpcContext.Provider>
-    )
+    return this._renderProviders(children)
   }
 }
 
-export function withVrpc (PassedComponent, mapVrpcToProps) {
+export function withVrpc (
+  backendsOrPassedComponent,
```

Comment:
Are we still sure we want to go with this function signature? I think it might be better to have it more explicit where the user has to specify they want everything by passing in null. Example below:

Current API:
```javascript
// Needs one backend
withVrpc('backendName', <ComponentThatNeedsOneBackend />)

// Needs all backends 
withVrpc(null, <ComponentThatNeedsAllBackends />)
```

Proposed API:
```javascript
// Needs one backend
withVrpc('backendName', <ComponentThatNeedsOneBackend />)

// Needs all backends 
withVrpc(null, <ComponentThatNeedsAllBackends />)
```

To me it feels like with passing null the user has to actively make the decision that they want everything and want the component to update for events. I feel like we need to either go with a more explicit API or warn the users in the documentation.

Reply:
If we specifically know which backends we need, can't we do `withVrpc(['backend1', 'backend2'], <ComponentThatNeedsBackends />`

Reply:
> If we specifically know which backends we need, can't we do `withVrpc(['backend1', 'backend2'], <ComponentThatNeedsBackends />`

Yes but thats not the point. The point is that if no backends are passed then all backends are provided to the component. I feel like there should be a specific opt into having all backends passed by passing in null or maybe even another withAllVrpc HOC. That or are users need to be very explicitly warned that if they pass all backends into each component they use they can face performance issues.

Reply:
@Benni2808 we can exactly do what you describe. The function is "overloaded", either no argument, a string or and array of strings are allowed.
@JacobTheEvans I do not like the `null` API too much. I could imaging that many people won't use several backends and this `null` then is boilerplate. Plus, we anyways have an overload between string and array of strings.

Reply:
> @Benni2808 we can exactly do what you describe. The function is "overloaded", either no argument, a string or and array of strings are allowed.
> @JacobTheEvans I do not like the `null` API too much. I could imaging that many people won't use several backends and this `null` then is boilerplate. Plus, we anyways have an overload between string and array of strings.

Alright Im willing to move forward with my approval but think we might want to discuss in the future after we have some more projects using this library.